### DISCUSSION TITLE
[Discussion] Discriminated Unions features "shopping list"

### DISCUSSION THREAD

**Discussion by DavidArno** [dotnet/csharplang#3760] [Language Ideas]
This is a write-up of my thoughts on what the "shape" of sum types (ie Discriminated Unions) in C# could look like. Throughout this article, I just refer to them as "unions".

In [Remaining design work in and around records](https://github.com/dotnet/csharplang/issues/3707), Mads mentions at the end that "*We've said that we'd look at discriminated union scenarios once we got records under control. That time has come. The design space is wide open at this point.*". This is an area that I've been thinking about for some time. As such I felt it now a good time to publish my effective "shopping list" of wants for unions in C#.

For the purposes of this article, I've invented syntax for how one might express union types. This is arbitrary syntax. I feel it consistent with the rest of the language, but the syntax used is unimportant. What I'm seeking to do here is express ideas around the functionality of unions and not to plant a flag in the ground of a particular syntax form.

For many of the examples here, I've borrowed heavily from [F# for fun and profit](https://fsharpforfunandprofit.com/posts/discriminated-unions/).

## A simple union ##

A union of an `int` and a `bool` would look something like this:
```c#
enum struct IntOrBool
{
    I { int },
    B { bool }
} 
```
Here, an instance of `IntOrBool` can either be an `I of int`, or it can be a `B of bool`. As such, pattern matching can be used to determine which it is. We can match against the 
```c#
var oneOfThem = new IntOrBool.I(2);
if (oneOfThem is I { var i })
{
    // it is an I in this case, and i is an int == 2 
}
```

`I` and `B` can be thought of as just being case labels, rather than definitely types themselves. Effectively how thy are handled by the compiler should be seen as an implementation detail.

## Shortcutting the labels for matching simple unions

In the example union above, both `I` and `B` are just wrappers over other single types. As such, the pattern match may be simplified to:
```c#
var oneOfThem = new IntOrBool.I(2);
if (oneOfThem is int i)
{
    // it is an I, which just wraps int and so the pattern matches 
}
```
The reasoning behind this is largely driven by the way nullable structs are handled by pattern matching. It seems desirable to have `int?` and `Option<int>` be handled the same when pattern matching them:
```c#
int? x = 1;
Option<int> y = 1;

if (x is int i1) ... // we don't express this as if (x is int? { int i1 }) ...
if (y is int i2) ... // therefore if (y is Some { int i2 })... seems burdensome in comparison to nullables
```
This could only apply where there was ambiguity over which case was being matched, so for:
```c#
enum Result<T>
{
    Success(T value),
    Error(string message),
}

var result = Result<string>.Error("message");
if (result is string value) ... // this would be an error as both Success and Error could match here
```

## Unions can be structs or classes ##

In the following example, `IntOrBool` is a struct and `IntOrString` a class. The usual rules of pass by value/reference would therefore apply to them, respectively:

```c#
enum struct IntOrBool
{
    I { int },
    B { bool }
} 

enum class IntOrString
{
    I { int },
    S { string }
} 
```

## Unions can be generic ##
For a union that can be either `T1` or `T2`:
```C#
enum struct Either<T1, T2>
{
    Left { T1 },
    Right { T2 }
}
```
This then highlights the benefit of the shortcutting approach to pattern matching:
```C#
var either = Either<int, bool>.Left(2);

if (either is Left { T1 value }) ...

if (either is T1 value) ...
```
"Left" and "Right" are just handy labels, but they carry no useful information and could just as easily be called "Sally" and "Harry". Removing them from the pattern match significantly increases the readability of the code.

## Constructing a union type ##
As seen in previous examples, one way of constructing a union is to fully express the type and label:
```c#
var foo = new IntOrBool.I(2);
```

If `I` is unambiguous then the type isn't needed:
```c#
var foo = new I(2); // foo is an instance of IntAndBool
```

If the type involves generics though, this doesn't work:
```C#
var either = Left(2); // Error. We could infer that T1 is int, but what's T2?
```

So, as long as the value is unambiguous with regards to the types in the union, the label isn't needed:
```C#
var either = Either<int, bool>(2); // It's a Left.
```

Finally, unions could support implicit casting:
```C#
Either<int, bool> either = 2;
```

## Empty cases ##

In the previous examples, the cases all have values of some type or other. But there are situations where a case doesn't have a meaningful value. The classic `Maybe`/`Option` type being one such example:
```C#
enum struct Option<T>
{
    None,
    Some { T }
}
```
In this case, pattern matching uses the label as normal:
```C#
var maybe = Some(2);

var result = maybe switch {
    None => "no value",
    int i => i.ToString();
}
```
And constructing a union instance with an empty case is just like using a parameterless constructor:
```c#
Option<int> maybe = None;
```

## Closed enums are just a union of empty cases ##

With regard to [Proposal: "Closed" enum types](https://github.com/dotnet/csharplang/issues/3179), a closed enum (ie one that has a fixed set of cases that cannot be extended at runtime) can be expressed via a union:
```c#
enum struct Bool { False, True }

int M(Bool b) => b switch {
    False => 0,
    True => 1
};
```
Unlike classic, "open" enums, the cases of such a closed enum do not have any values so cannot be cast to/from eg `int`.

## Single cases ##

So far, all the examples have been of unions with at least two cases. However, single case unions can be used to address [the primitive obsession problem](https://refactoring.guru/smells/primitive-obsession) without the need for lots of boilerplate code.

```c#
enum class PostCode { PostCode { string } }

enum class TelephoneNumber { Number { string } }

void HandlePostCode(PostCode) => ...

PostCode postCode = "AA1 1AA"; // string can be cast to PostCode
TelephoneNumber phone = "001 555 ...";

HandllePostCode(postCode); // valid
HandllePostCode("AA1 1AA"); // Also valid as string can be cast to PostCode
HandlePostCode(phone); // invalid
```

And the value from such a single case can be "deconstructed" with pattern matching as required:
```c#
if (postCode is string s)
{
    ...
}
```

## Cases can have values of the same type ##
```c#
enum struct Telephone
{
    Landline { string },
    Mobile { string }
}
```
In such cases, the label must be explicitly specified during construction:
```c#
Telephone phone = new Telephone("12345"); // error as this could be a LandLine or Mobile

var mobile = new Mobile("077..."); // All good
```

And the label is also needed in pattern matching:
```c#
if (mobile is Mobile { var s }) ...
```

## Cases can have more than one value and the values can be named ##
Up until now, all the examples have just used one value, which has not had a name and so a variation on recursive pattern matching has been used to access the value:
```c#
if (mobile is Mobile { var s }) ...
```

But sometimes we'd want to give it a name. Also sometimes we'd want that label to take more than one value, in which case they definitely want names. An example might be:

```c#
enum struct Shapes 
{
    Point,  
    Rectangle { double width, double length },
    Circle { double radius }
}
```
`Circle` doesn't need the name, `radius`. But it helps clarify what the value is. A `Rectangle { double, double }` is pretty meaningless by itself though, so it makes sense to name them. 

With names, ordinary recursive pattern matching can be used:
```c#
var area = shape switch {
    Point => 0,
    Rectangle { width: var w, length: var l } => w * l,
    Circle { radius: var r } => Math.PI * r * r
};
```
though the `radius:` part is optional.

## Union equality ##
For both class-based and struct-based unions, equality would be value-based. Copying the equality rules for records likely makes sense. As those rules (especially around `==`) have the potential to change, I'll not detail equality further, therefore.

## Recursive cases ##
As C# types already support self-referencing to create recursive types, this can be applied to unions too. For example, a generic  tree structure could be expressed as follows:
```c#
enum class Tree<TLeafData, TNodeData>
{
    LeafNode { TLeafData },
    InternalNode { IEnumerable<(TNodeData, Tree<TLeafData,TNodeData>)> }
}
```

## Serialization, deconstruction and `ToString()` ##
Unions will need to support serialization/deserialization somehow. The challenge is that many serialization frameworks rely on property-based serialization. This doesn't fit nicely with unions as far as I can see. I therefore don't have a proposed solution to this at this time.

Deconstructors for unions do not make much sense to me as C# does not support multiple deconstructors with the same arity and difference parameter types. Therefore my feeling is that a union shouldn't auto-implement a `Deconstruct` method.

The default `ToString` for a union likely can just be of the form:
```
<Label> [[<Name>:]<Value>(,[<Name>:]<Value>)*]
```
So in the `Shapes` example above, a rectangle of `10.0 x 20.0` would result in:  
`Rectangle width:10.0,length:20.0`.

## User-defined code within unions ##

In the case of single case unions, as well as preventing muddling primitives and supplying metadata around the purpose of the value, it's common to want to support value validation too. So allowing user-defined code, such as supplying a validating constructor, should be considered for unions. 

## Inheritance ##

As unions are a closed set of types, class-based unions should be sealed and thus cannot be inherited from.

There are likely use cases where it makes sense for class-based unions to inherit from a base class though.

**Comment by 333fred** [comment]
This is a very good writeup. You've hit a lot of points that I think are important to me as well, and I'll certainly keep this issue in mind as we move forward with design.

One small comment, I think you forgot an `is` in the shortcutting section? Fwiw that's one of the sections I'm more iffy on: it could be considered unambiguous, but that feels like too much magic to me, particularly as introducing a new case could be a source breaking change. It seems like too much spooky action at a distance.

**Comment by HaloFour** [comment]
Debating syntax as always is bike-shedding but I envision DUs to be more built on top of record syntax.  So given you can define records with the following syntax:

```cs
public record Rectangle(int Length, int Width);
public record Circle(int Radius);
```

You could define a DU as (something like) the following:

```cs
public enum Shape {
    Rectangle(int Length, int Width),
    Circle(int Radius)
}
```

What that actually generates if certainly up for a lot of debate.

**Comment by HaloFour** [comment]
I have a feeling that DUs will likely end up meaning more than one thing depending on how they're intended to be used.

You could have closed type hierarchies where each of the cases is a separate type that inherits/implements from a base type, such as how Java and Scala 2.x approach ADTs:

```java
package com.example.expression;

public sealed interface Expr
    permits ConstantExpr, PlusExpr, TimesExpr, NegExpr {...}

public record ConstantExpr(int i)       implements Expr {...}
public record PlusExpr(Expr a, Expr b)  implements Expr {...}
public record TimesExpr(Expr a, Expr b) implements Expr {...}
public record NegExpr(Expr e)           implements Expr {...}
```

And Scala 3.x introduces a nice shorthand for that:

```scala
enum Option[+T] {
  case Some(x: T)
  case None
}
```

You could also have tagged unions where a type hierarchy wouldn't work, such as with structs where you'd want a zero- or low-cost abstraction.  This is how F# implements them:

```fsharp
[<Struct>]
type Option<'a> =
    | Some of 'a
    | None
```

Maybe a single implementation can offer the flexibility of a type hierarchy with the performance characteristics of a tagged union?

**Comment by jnm2** [comment]
> "Left" and "Right" are just handy labels, but they carry no useful information and could just as easily be called "Sally" and "Harry". Removing them from the pattern match significantly increases the readability of the code.

I'm okay with this as syntax sugar for using `Left` and `Right` if T1 and T2 are different, but there is a big problem for me if matching is actually type-based and not actually nominal. One of my most commonly-used DUs, in pseudocode:

```cs
enum Result<T>
{
    Success(T value),
    Error(string message),
}
```

What should happen here?

```cs
var result = Result<string>.Error("message");
if (result is string value)
{
    // Is this Success or Error?
```

Side note: I would love for there to be a way to use the syntax `Result.Error("message")` and obtain an instance that implicitly casts to any `Result<T>`, but I don't know of any way to generalize this. What I currently do is to define `Error` on a static nongeneric `Result` class and return a `Result.ErrorResult` struct which `Result<T>` declares an implicit conversion from.

**Comment by HaloFour** [comment]
@jnm2 

I agree, I personally think that the case names are more important than the types of their elements, especially since they could be a singleton (e.g. `None`) or have any number of elements.  I don't think allowing them to be elided makes sense in the match anymore would it would to elide the property names in a recursive property pattern.

Even with a simple ADT where the element should be obvious I think it makes more sense to match based on the name of the case and use recursive patterns to extract the elements:

```cs
// use this
if (opt is Some(var value)) { ... }

// not this
if (opt is string value) { ... }
```

**Comment by Unknown6656** [comment]
@HaloFour 
> Debating syntax as always is bike-shedding but I envision DUs to be more built on top of record syntax. So given you can define records with the following syntax:

~We have to be careful with that syntax. It could be easily confused with a method declaration. What about using the already existing `case`-keyword for union case declarations?~

------------
@DavidArno One important point: "anonymous" unions (syntax is debatable):
```csharp
(union Option<int>{ Some { int }; None }) TryParseInt32(string s) => .....;
```
I'd hate to declare a `OptionalInt`-union for this case. The solution should be analogous to current tuple usage.

**Comment by TheUnlocked** [comment]
To add to the discussion on single-type cases, rather than allowing you to match on the type like
```cs
if (opt is T1 value) { ... }
```
Or forcing you to match like
```cs
if (opt is Some { var value }) { ... }
```
I'd prefer allowing
```cs
if (opt is Some value) { ... }
```
It's more descriptive, just as concise, and probably less prone to errors.

_Sorry if this is edging too far into bikeshed territory >.<_

**Comment by 333fred** [comment]
```cs
if (opt is Some value) { ... }
```
This already means something today.

**Comment by TheUnlocked** [comment]
Ah...-- if you actually want `value` to stay as an `Option` my proposed alternative would be an issue. Alright, I concede there.

**Comment by FiniteReality** [comment]
I personally believe eliding case names for discriminated unions would be far too dangerous, leading to code that actually ends up being far more brittle. (Especially in the `Result<string>` case demonstrated with an Error case which contains a string for the error message.)

Other than that, I don't really see any issues with the features on your wishlist.

**Comment by AartBluestoke** [comment]
even in the non-generic case the type can be ambiguous - i deal with Microsoft Dynamics and Entities can be owned by any Identity Principal, and at the moment i have to manually track or ask what type of guid i'm dealing with:

enum mscrm.principal{
systemuser {Guid},
team {Guid},
businessUnit {Guid}
}

**Comment by orthoxerox** [comment]
> In the following example, IntOrBool is a struct and IntOrString a class. 

Is there a reason why one has a bool and the other a string? If there's none, then this can lead the reader astray.

> If the type involves generics though, this doesn't work:

It should still work if we're not using local type inference, `return Left(2);` should pick up the generic types from the method signature.

**Comment by ghost** [comment]
> ## Single cases
> ...
> 
> And the value from such a single case can be "deconstructed" with pattern matching as required:
> 
> ```cs
> if (postCode is string s)
> {
>     ...
> }
> ```
> 
Having to use if or switch for the infallible extraction of the underlying value of a "type-wrapper" union would be inconvenient:
```cs
if (!(contact.Name is string name) || !(contact.EMail is string mail) || !(contact.PhoneNumber is string phone))
{
    throw new UnreachableException();
}
string contactBlock = $"{name}\nM: {mail}\nP: {phone}";
```
We could add shorter syntax for these infallible cases, for example cast-style:
```cs
string contactBlock = $"{(string)contact.Name}\nM: {(string)contact.EMail}\nP: {(string)contact.PhoneNumber}";
```
But if we would create syntax for that use-case anyway, I'd say it would make more sense to not abuse unions for that and implement a specialized syntax like #1695.

**Comment by DavidArno** [comment]
@333fred, typo fixed. Thanks.

With regard to the comments from yourself and others regarding the "shortcutting section", I've expanded that section to explain my thinking further. The key point for me is that I'd like to see syntax parity in matching eg `int?` and `Option<int>`:
```c#
int? x = 1;
Option<int> y = 1;

if (x is int i1) ... // we don't express this as: if (x is int? { int i1 }) ...
if (y is int i2) ... // therefore if (y is Some { int i2 })... seems burdensome in comparison to nullables
```

With regards to @jnm2's example, I'd see this only applying though if there were no ambiguity over which case was being made:
```c#
enum Result<T>
{
    Success(T value),
    Error(string message),
}

var result = Result<string>.Error("message");
if (result is string value) ... // this would be an error as both Success and Error could match here
```

But I agree that there is a degree of "magic" to this and maybe that's too high a price to pay to simplify just one edge-case union (`Option`/`Maybe`).

**Comment by HaloFour** [comment]
@DavidArno 

> The key point for me is that I'd like to see syntax parity in matching eg `int?` and `Option<int>`.

I can see wanting that equivalence but I don't know that it makes sense to affect all other DUs for that special case.  And without additional runtime support it would not behave like `Nullable<T>` when boxed and that pattern would never match.  IMO, it'd be sufficient to use a pattern like `Some(var x)` to match and extract the value from `Option`/`Maybe`.

When it comes to DUs like `Option` what concerns me more is the cost of the abstraction and the ergonomics of matching if each "case" isn't a separate type.

**Comment by orthoxerox** [comment]
One more thing: struct-based DUs will benefit from non-defaultable value types ([latest discussion](https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-07-01.md)) a lot. While some DUs can be arranged so that their semantically default value matches the zero-bit pattern, some (like Either) don't have a default value to speak of.

**Comment by themeeman** [comment]
I'm not a huge fan of the implicit conversions, even in the cases where its unambiguous. I think you should always have to specify the name of the alternative of the union you are calling, since otherwise adding new alternatives could break existing code. This problem could be solved by adding an implicit conversion operator, which would require the person building the type to opt-in.

**Comment by stephen-hawley** [comment]
Multiple things - why not make the syntax a little more clear that this is a union:
```
enum struct IntOrBool {
    int I | bool B
}
```
This makes it more clear that it's choice `I` *or* `B` It also tightens things up a bit.
What does `typeof` return? I assume `IntOrBool`. Will there be a `caseof` pseudo function? It would be handy to also generate an enum maybe under the same name such that `IntOrBool.I` is the value of the descriminator and `IntOrBool.I(n)` is the factory.

If this is a true DU, then the payloads in the actual implementation should overlap so as to not waste memory. This is an issue from a garbage collection standpoint, however. How will that be handled?
I assume also that I can have methods on these as well.

**Comment by jnm2** [comment]
@stephen-hawley What about `I(int x, int y)`?

**Comment by FiniteReality** [comment]
> Multiple things - why not make the syntax a little more clear that this is a union:
> 
> ```
> enum struct IntOrBool {
>     int I | bool B
> }
> ```

I personally disagree with this syntax... It doesn't really do much to clarify that it's a union or clarify that there's a choice IMHO. Since we're clearly into bikeshedding category, I'd like to see something akin to the Rust syntax, which I feel really did get discriminated unions more or less right:

```rust
enum Animal {
    Dog(String, f64),
    Cat { name: String, weight: f64 },
}

let mut a: Animal = Animal::Dog("Cocoa".to_string(), 37.2);
a = Animal::Cat { name: "Spotty".to_string(), weight: 2.7 };
```
(This example is from their reference on [enums](https://doc.rust-lang.org/reference/items/enumerations.html))

This is much like the syntax examples given by the OP, which I feel fits very much in line with C# as a language.

**Comment by Unknown6656** [comment]
> This is much like the syntax examples given by the OP, which I feel fits very much in line with C# as a language.

Agreed, however, @stephen-hawley's view point probably comes from the "traditional" syntax of describing union types (e.g. in ~Haskell,~ OCaml, F#, etc.):
```fsharp
type IntOrBool = 
    | I of int
    | B of bool
```
I do love the usage of `|` to represent a choice, as it has some symmetry with the traditional "Or"-operator.
This, however, is not a smart syntax for languages such as C# which are not indentation-sensitive.

**Comment by orthoxerox** [comment]
Some DU syntax sugar shopping list:

**C# 9.0**

First, an example of current pain points:

```c#
//Poor man's DU:
public abstract record Result<T, TError>
{
    public record OK(T ok) : Result<T, TError>;
    public record Error(TError error) : Result<T, TError>;

    public static implicit operator Result<T, TError>(T ok) => new OK(ok);
    public static implicit operator Result<T, TError>(TError error) => new Error(error);
}
```

```c#
public record ConditionalExpression(Expression Condition, Expression IfTrue, Expression IfFalse) : Expression
{
    public override Result<Type, TypeError> DetermineType()
    {
        var condType = Condition.DetermineType();
        
        switch (condType) {
            case Result<Type, TypeError>.OK(var ct) when ct != ScalarType.Boolean:
                return new TypeError(Condition, $"Expected boolean, got {condType}");
            case Result<Type, TypeError>.Error err:
                return err;
        }
        
        if (condType != ScalarType.Boolean) {
            return new TypeError(Condition, $"Expected boolean, got {condType}");
        }

        var trueType = IfTrue.DetermineType(); //If you see below, we just return the errors
        var falseType = IfFalse.DetermineType();

        //Every time we have to repeat Result<Type, TError>
        return (trueType, falseType) switch {
            (Result<Type, TypeError>.OK(var t), Result<Type, TypeError>.OK(var f)) => Type.FindCommonType(t, f) switch {
                Result<Type, TypeError>.OK ok => ok,
                Result<Type, TypeError>.Error(var err) => new TypeError(this, err.Message),
            },
            (Result<Type, TypeError>.Error err, _) => err,
            (_, Result<Type, TypeError>.Error err) => err,
        };
    }
}
```

**C# 10**

**Sugar A: No need to specify the DU type, just the case names**

```c#
//Real DU as specced in https://github.com/dotnet/csharplang/blob/master/proposals/discriminated-unions.md
public enum class Result<T, TError>
{
    OK(T ok),
    Error(TError error),
}
```

```c#
        switch (condType) {
            case OK(var ct) when ct != ScalarType.Boolean:
                return new Error(new TypeError(Condition, $"Expected boolean, got {condType}"));
            case Error err:
                return err;
        }
```

```c#
        return (trueType, falseType) switch {
            (OK(var t), OK(var f)) => Type.FindCommonType(t, f) switch {
                OK ok => ok,
                Error(var err) => new Error(new TypeError(this, err.Message)),
            },
            (Error err, _) => err,
            (_, Error err) => err,
        };
```

This is *much* more legible now. The implicit conversion is gone now, increasing the verbosity slightly, but that's outside the scope of this comment.

**Sugar B: An unwrap-or-bailout sugar that returns the failure case immediately, unpacking the success case otherwise**

One specific implementation is proposed in #3403, but this is about the sugar, not the implementation specifics. Yes, I am aware that this applies only to a couple DU's.

```c#
public record ConditionalExpression(Expression Condition, Expression IfTrue, Expression IfFalse) : Expression
{
    public override Result<Type, TypeError> DetermineType()
    {
        var condType = Condition.DetermineType()🚀;
        if (condType != ScalarType.Boolean) {
            return new Error(new TypeError(Condition, $"Expected boolean, got {condType}"));
        }

        var trueType = IfTrue.DetermineType()🚀; //The rocket is just a placeholder symbol for the real sugar
        var falseType = IfFalse.DetermineType()🚀;
        //Now the types of trueType and falseType are no longer Result<Type, TypeError>, but Type

        //This allows us to skip the outer match
        return Type.FindCommonType(trueType, falseType) switch {
            OK ok => ok,
            Error(var err) => new Error(new TypeError(this, err.Message)),
        };		
    }
}
```

The method is dramatically more legible now, in my opinion. Implicit conversions would help a lot to remove the final clunkiness, but that can be solved by allowing methods inside DU definitions.

**Comment by foxesknow** [comment]
Can we not introduce a new keyword `union`:


```
union struct Expression{}
union class Expression{}
```


The `record` type was added to better document the intent, so why not do the same for discriminated unions?

**Reply by FiniteReality** [reply]
Discriminated unions are closer to enumerations than they are true unions.

Plus, C# has a large influence from predecessor languages like C and C++, where `union` is already an existing thing - this would likely be more confusing.

**Reply by HaloFour** [reply]
I agree, and `enum` is already the keyword of choice in many languages that have discriminated unions, like Swift, Rust, Haxe and Scala 3.0 (aka Dotty).  Enumerations in C# and Java can already be considered as a form of discriminated union of unit types.

**Comment by stephen-hawley** [comment]
Love this.
For the problem domain that I'm looking at (modeling Swift discriminated unions at run time), it is **vital** that in addition to syntax, it needs to be possible to explicitly break out the discriminator and the payload accessors. The syntax is immaterial to me, but to put a fork in the road, let's assume that when you define a DU you there is an implicit closed enum defined named Cases and accessors for them:
```
enum class SwiftEnum {
    enum Cases {
        First = 0,
        Second = 1
    }
    // private member
    byte[] swiftPayload;
    First { FirstSwiftTypeSurfacedInCSharp } {
        get => return pInvokeToUnpackFirstPayload(swiftPayload); // need more actual marshaling
        set =>  pInvokeToMutateFirstPayload (swiftPayload, value);
    }
    Second { SecondSwiftTypeSurfacedInCSharp } {
        get => return pInvokeToUnpackSecondPayload(swiftPayload); 
        set =>  pInvokeToMutateSecondPayload (swiftPayload, value);
    }
    Cases Discriminator => return pInvokeToUnpackDiscriminator(swiftPayload);    
}
```

Reasoning: in modeling swift enums (DU's), there is an non-trivial process to determine and extract the payload of an enum. This process is dependent on Swift's ABI and is non-trivial. As such it is best left to Swift. By breaking out the accessors as methods that can be provided __optionally__ then these types can be modeled in C# cleanly.

This can be extended beyond Swift - this applies to any language that needs to interoperate with C#.

**Comment by Richiban** [comment]
Re: syntax, I'd like to make the case (a ha.) for using the keyword `case` for the enumerated types:

```cs
enum class PaymentMethod
{
    case AccountCredit;
    case CreditCard(string CardNumber);
    case PayPal(string EmailAddress);
}
```

there can be a shorthand for this: 

```cs
enum class Colour
{
	case Red, Green, Blue
}
```

The reason I'd like the extra keyword for each of the enum cases is that, since this is an enum **class**, we should be able to put methods and other members in the base, like so:

```cs
enum class Shape
{	
	case Point;
	
	case Line(int Length);
	
	case Rectangle(int Width, int Height)
	{
		public override double Area => Width * Height;
	}
	
	public virtual double Area => 0;
}
```

If we don't have a keyword before every case I think it's going to be syntactically hard to distinguish members from cases.

**Reply by HaloFour** [reply]
Your syntax seems to suggest that the `enum class` type would be represented by a class hierarchy and my concern is if/how that would translate to `enum struct` types, unless `virtual`/`override` members are forbidden or interpreted in a completely different way.  Also, how would the simple cases like `Point` be represented?

I kinda like requiring `case`, feels familiar coming from Scala, but IMO those syntax details are bikeshedding compared to the rest of the design in this space.  😜

**Reply by Richiban** [reply]
I had hoped that I was clear I want to avoid bikeshedding here; I didn't mean to talk specific syntax except to argue that enum classes should still support most of the features of regular classes, and we therefore need some way of differentiating enum cases from regular class members. The actual syntax isn't important to me right now.

I'd always imagined that `enum structs` would (obviously) function completely differently, due to the lack of any inheritance. Maybe something clever can be done to minimize their size through FieldOffset in a way that doesn't trigger TypeLoadExceptions at runtime. I don't believe that what I'm talking about here will have any bearing on `enum struct`s, should they arrive later.

'Value' cases such as `Point` in my example would be implemented as a singleton class (which is currently a concept that doesn't exist in C#, at least not at the language level). My example would essentially lower to:

```cs
[ClosedType]
abstract record Shape
{
	private Shape() { }

	public record Point : Shape
	{
		private Point() { }

		public static readonly Point Instance = new();
	}

	public record Line(int Length) : Shape;

	public record Rectangle(int Width, int Height) : Shape
	{
		public override double Area => Width * Height;
	}

	public virtual double Area => 0;
}
```

There would need to be some more compiler sugar to hide all the necessary usages of `.Instance` but I don't think that's too hard. To be honest, I'd like it if C# would allow top-level `object` declarations, like Scala and Kotlin, that are a type declaration that cannot be instantiated but it has a singleton value that you can reference just with the type name.

**Comment by stephen-hawley** [comment]
Looking at this from my project's point of view (modeling Swift types), C# discriminated unions (DU) are useful for me if and only if:

  1.  Behave properly with automatic reference counting (ARC)
  2.  Code can be provided to implement discriminator and payload methods

With regards to the latter, imagine that the actual implementation of a DU includes an interface like this:
public IDiscriminatedUnion<D> where D: System.Enum {
    D Discriminator { get; }
    object Payload { get; }
    static Type PayloadType (D discriminator);
}

That way it's possible to create a partially de-sugared implementation of the enum, which would allow me to model Swift DU's more directly.

Longer explanation of the need:
The swift DU implementation is...inspired. The compiler performs a step where it looks at the union of all types in the DU and looks for "extra inhabitants" these are bits that are common to all types that aren't part of the type. For example, on some platforms the top and/or bottom bits of pointers aren't used. If the log2(case count) < number of extra inhabitants, they use the extra inhabitants. Otherwise it will fall back on a number of other strategies. Plus the strategy chosen may be affected by the compiler settings (--enable-library-evolution for example). Because of this, there is no direct way to find out the discriminator of a Swift DU without writing code in Swift.

As for the need for a deinitializer. Suppose I have a modeled Swift enum using the above but without a deinitializer, and I define an enum like this:
public SomeSwiftEnum : IDiscriminatedUnion<SomeSwiftEnum> {
    case Foo(SomeSwiftHeapType), case Bar(nint);
    byte [] payload;
    public unsafe SomeSwiftEnum Discriminator => DiscPinvokeIntoSwift(&payload);
    public object Payload => MarshalToCSharp(PayloadType, payload);
    public static Type PayloadType(SomeSwiftEnum e) {
        if (e == SomeSwiftEnum.Foo) return typeof (SomeHeapType);
        return typeof (nint);
    }
}

So I use this type in this context:
{
     var en = SomeSwiftEnum(new SomeSwiftType());
     DoSomeWork(en); // whatever
} // possible memory leak because the reference count on the underlying handle in the heap object is floating

I mean maybe this will work because the heap object is IDisposable and the garbage collector should  pick that up? But still. There's also the issue that this particular enum isn't blittable and copying it into an array of SomeSwiftEnum presents other issues.

________________________________

**Reply by HaloFour** [reply]
Those requirements seem incredibly specific to trying to model Swift DUs and don't seem like they would work from a general purpose point of view.  For starters, GC is a runtime concern, and the .NET runtime doesn't do ARC, so you'd have to forego that requirement out of the gate.

**Comment by stephen-hawley** [comment]
Understood, but the approach of providing the tools to implement a full or partially desugared version of the type make it much more flexible. Kind of like being able to  use the elements of linq without the syntax. I think if the GC properly understands how to manage an IDiscriminatedUnion<T> (and it will have to), then I can use my existing model. Plus this approach allows interop beyond my particular niche.
I can be selfish and selfless at the same time. 🙂


Sent from Outlook<http://aka.ms/weboutlook>

________________________________

**Comment by erik-kallen** [comment]
Nice write-up.

I'm not a big fan of the shortcuts for matching on just a member or implicit conversions, though. I realize what the use case is about, but I think it would be better if the classes define some methods for this, somehow, so we'd have something like

```
enum class Option<T> {
    Some { T },
    None,
    public static implicit operator Option(T value) => Some(value)
    public static bool Deconstruct(Option<T> option, out T value)  {
        if (option is Some { var v }) {
            value = v;
            return true;
        }
        else {
            value = null;
            return false;
        }
    }
}
```

**Comment by NielsHoogeveen** [comment]
> ## A simple union
> A union of an `int` and a `bool` would look something like this:
> 
> ```cs
> enum struct IntOrBool
> {
>     I { int },
>     B { bool }
> } 
> ```
> 
> Here, an instance of `IntOrBool` can either be an `I of int`, or it can be a `B of bool`. As such, pattern matching can be used to determine which it is. We can match against the
> 
> ```cs
> var oneOfThem = new IntOrBool.I(2);
> if (oneOfThem is I { var i })
> {
>     // it is an I in this case, and i is an int == 2 
> }
> ```
> 
> `I` and `B` can be thought of as just being case labels, rather than definitely types themselves. Effectively how thy are handled by the compiler should be seen as an implementation detail.
> 
I think the construction of a union of int and bool should not be defined upfront, but should be defined upon its usage.
It could for example be written like this:

````cs
public void DoSomethingWithIntOrBool(int | bool value){
   value match{
    case int i => // do int stuff
    case bool b => // do bool stuff
   }
}
````
We could go fancy and define the signature of the sum type, and use that at the call site:
````cs
type IntOrBool = int | bool
public void DoSomethingWithIntOrBool(IntOrBool value){
   value match{
    case int i => // do int stuff
    case bool b => // do bool stuff
   }
}
````
There should not be a need to wrap a sum type into a union for those cases where the types already exist and cannot be changed or extended.

**Reply by csharper2010** [reply]
You have a point regarding that example.

But generally, we should not mix union types with discriminated unions. I am more desperately awaiting the latter.

**Comment by ziaulhasanhamim** [comment]
Will C# really have Discriminated Unions ever? This is one of the most wanted features. But still, progress is very slow on Discriminated Unions

**Reply by Unknown6656** [reply]
@MadsTorgersen What are the perquisites / hurdles to overcome in order to bring DUs into C#?
Are runtime changes needed? Is the LDT unsure concerning the syntax or implementation/compiler generation?
What is the current progress on that front?
